doc/debian-distro-info.1
.\" Copyright (c) 2009-2014, Benjamin Drung <bdrung@debian.org>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.TH DEBIAN\-DISTRO\-INFO "1" "August 2013" "distro\-info"
.SH NAME
debian\-distro\-info \- provides information about Debian's distributions
.SH SYNOPSIS
.B debian\-distro\-info
[\fIOPTIONS\fR]
.SH OPTIONS
.TP
\fB\-\-date\fR=\fIDATE
date for calculating the version (default: epoch taken from SOURCE_DATE_EPOCH
environment variable if set, otherwise today)
.TP
\fB\-h\fR, \fB\-\-help\fR
display help message and exit
.TP
\fB\-\-alias\fR=\fIDIST
print the alias (oldstable, stable, testing, unstable) relative to
the distribution codename passed as an argument.
Only distribution codenames composed of lower case ASCII letters are accepted,
and if the distribution does not qualify as oldstable, stable, testing, or
unstable, then the same codename passed as argument is returned.
.TP
\fB\-a\fR, \fB\-\-all\fR
list all known versions
.TP
\fB\-y\fR[\fIMILESTONE\fR], \fB\-\-days\fR[=\fIMILESTONE\fR]
display number of days until specified version reaches the specified milestone.
.I MILESTONE
may be one of
.IR created ", "
.IR release ", or "
.IR eol "."
If no milestone is specified, assume \fIrelease\fP.
For options that return a list, display the normal output followed by
whitespace and the number of days until the specified milestone.
If milestone cannot be calculated, the string \(aq(unknown)\(aq is displayed.
Unless one of \fB\-c\fP, \fB\-f\fP or \fB\-r\fP is specified,
only the number of days will be displayed.
.TP
\fB\-d\fR, \fB\-\-devel\fR
latest development version
.TP
\fB\-o\fR, \fB\-\-old\fR
latest old (stable) version
.TP
\fB\-\-series\fR=\fISERIES
series to calculate the version for
.TP
\fB\-s\fR, \fB\-\-stable\fR
latest stable version
.TP
\fB\-\-supported\fR
list of all supported versions, including development releases
.TP
\fB\-t\fR, \fB\-\-testing\fR
latest testing version
.TP
\fB\-\-unsupported\fR
list of all unsupported stable versions
.TP
\fB\-c\fR, \fB\-\-codename\fR
print the codename (default)
.TP
\fB\-r\fR, \fB\-\-release\fR
print the release version
.TP
\fB\-f\fR, \fB\-\-fullname\fR
print the full name
.SH SEE ALSO
.BR distro\-info (1),
.BR ubuntu\-distro\-info (1)
.SH AUTHOR
The script and this manual page was written by
Benjamin Drung <bdrung@debian.org>.


debian/changelog
distro-info (0.21+deb10u1) buster-security; urgency=medium

  * Non-security LTS update, enabling distro-info-data updates:
    - Handle currently unknown columns in CSV data.
    - Make the Python module version PEP440 compliant.
    - Update tests for distro-info-data 41+deb10u8, which adjusted Ubuntu
      10.04 and 12.04's EoL
    - Change test unknown eol-server dates to use Ubuntu 12.10, which won't
      get ESM data in the future, if we include extra distro-info-data
      columns.

 -- Stefano Rivera <stefanor@debian.org>  Mon, 30 Oct 2023 11:39:24 +0200

distro-info (0.21) unstable; urgency=medium

  [ Benjamin Drung ]
  * Fix spelling mistake of exactly.
    Thanks to Brian Murray <brian@ubuntu.com> (Closes: #921581)
  * Bump Standards-Version to 4.3.0 (no changes needed)

  [ Balint Reczey ]
  * Support getting release version by codename in the Perl and Python API
    (Closes: #896640)

 -- Benjamin Drung <bdrung@debian.org>  Sun, 10 Feb 2019 00:29:24 +0100

distro-info (0.20) unstable; urgency=medium

  * Fix build failure with LC_ALL=POSIX by explicitly opening
    debian/changelog as UTF-8 in setup.py.

 -- Benjamin Drung <bdrung@debian.org>  Fri, 12 Oct 2018 12:00:35 +0200

distro-info (0.19) unstable; urgency=medium

  [ Stefano Rivera ]
  * Migrate VCS to salsa.debian.org.

  [ Ondřej Nový ]
  * d/control: Remove ancient X-Python-Version field
  * d/control: Remove ancient X-Python3-Version field

  [ Iain Lane ]
  * Fix pylint issues (Closes: #909987):
    + Allow useless-object-inheritance so we can stay PY2/3 bilingual.
      - This is correctly an unknown option when linting python 2, so we need
        to whitelist unknown overrides too.
    + Refactor a useless if/return/elif... chain into a sequence of `if`s
      instead.
    + Use chained comparison x <= y <= z instead of x <= y and y <= z.

  [ Benjamin Drung ]
  * Update pylint/flake8 unittests to latest version from
    https://github.com/bdrung/snippets
  * Support running tests against installed binaries and libraries
  * Add autopkgtest test suite
  * Set Rules-Requires-Root: no
  * Bump Standards-Version to 4.2.1
  * Do not install distro_info_test into Python libraries
  * Do not install test-*-distro-info scripts

 -- Benjamin Drung <bdrung@debian.org>  Fri, 12 Oct 2018 01:30:13 +0200

distro-info (0.18) unstable; urgency=medium

  * Update Python unittests to support pylint 1.8 (Closes: #889537)
  * Bump debhelper to version 11
  * Bump Standards-Version to 4.1.3 (no changes needed)
  * Update copyright year and format spec URL to https

 -- Benjamin Drung <bdrung@debian.org>  Sun, 04 Feb 2018 20:42:26 +0100

distro-info (0.17) unstable; urgency=medium

  * python: Strip trailing summary from pylint output
    (to fix build failure on Ubuntu 17.10)

 -- Benjamin Drung <bdrung@debian.org>  Fri, 21 Jul 2017 17:49:07 +0200

distro-info (0.16) unstable; urgency=medium

  * python: Introduce DistroRelease object and support returning it
    (steps to fix LP #1253208)
  * Add dh-python to build dependencies
  * Bump Standards-Version to 4.0.0 (no changes needed)

 -- Benjamin Drung <bdrung@debian.org>  Thu, 20 Jul 2017 12:08:34 +0200

distro-info (0.15) unstable; urgency=medium

  [ Mike Miller ]
  * debian-distro-info: Support oldstable with --alias. (Closes: #784284)

  [ Benjamin Drung ]
  * Fix flake8 and pylint issues found in Python code
  * Run flake8 and pylint checks on Python code in test suite
  * Python: Add DistroInfo.get_all() method
  * Bump Standard-Version to 3.9.8 (no changes required)
  * Use https for Vcs-Git and Vcs-Browser URLs

 -- Benjamin Drung <bdrung@debian.org>  Fri, 28 Apr 2017 19:45:36 +0200

distro-info (0.14) unstable; urgency=medium

  * Fix days in month calculation for leap years (Closes: #766142)
  * Bump Standards-Version to 3.9.6

 -- Benjamin Drung <bdrung@debian.org>  Tue, 21 Oct 2014 10:09:57 +0200

distro-info (0.13) unstable; urgency=medium

  [ Benjamin Drung ]
  * ubuntu-distro-info: Add --latest option. Use this option instead of --devel
    if you want to fallback to --stable for outdated data. (LP: #1309591)
  * distro_info_test/test_help.py: Close file descriptors at the end.

  [ Dimitri John Ledkov ]
  * Add --series lookup.

 -- Benjamin Drung <bdrung@debian.org>  Sat, 10 May 2014 20:55:33 +0200

distro-info (0.12) unstable; urgency=medium

  * Fix Ubuntu EOL test cases (The EOL date of raring is two days later).
    (Closes: #735801)
  * Bump Standards-Version to 3.9.5.

 -- Benjamin Drung <bdrung@debian.org>  Sat, 18 Jan 2014 11:45:43 +0100

distro-info (0.11) unstable; urgency=low

  * Added ability to specify --days[=MILESTONE] option.
    Thanks to James Hunt for implementing it.
  * Use canonical URLs for Vcs fields.
  * Bump Standards-Version to 3.9.4 (no changes needed).
  * Switch to debhelper 9.

 -- Benjamin Drung <bdrung@debian.org>  Fri, 02 Aug 2013 23:35:40 +0200

distro-info (0.10) unstable; urgency=low

  * Provide more useful errors (pointing to a README in distro-info-data) when
    the data is outdated.
  * Actually test with Python 3 (typo)
  * libdistro-info-perl: Print a warning and return 0 when data is out of
    date, rather than dying.
  * Use dpkg-buildflags and enable hardening.

 -- Stefano Rivera <stefanor@debian.org>  Fri, 01 Jun 2012 18:36:12 +0200

distro-info (0.9) unstable; urgency=low

  * debian-distro-info: Add an --alias option. Thanks to Antonio Ospite
    for the work. (Closes: #645171)
  * Make Python code valid Python 2 and 3 code and add a Python 3 module.
  * Fix one memmory leak in main() function.

 -- Benjamin Drung <bdrung@debian.org>  Fri, 13 Apr 2012 16:02:40 +0200

distro-info (0.8.2) unstable; urgency=low

  * Drop liblist-compare-perl from Build-Depends. Use our own list comparison
    functions in the test suite.

 -- Stefano Rivera <stefanor@debian.org>  Thu, 05 Apr 2012 11:39:55 +0200

distro-info (0.8.1) unstable; urgency=low

  * Fix build failure due to wrong variable type.


debian-distro-info.c
/*
 * Copyright (C) 2012-2013, Benjamin Drung <bdrung@debian.org>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#define DEBIAN
#define CSV_NAME "debian"
#define CSV_HEADER "version,codename,series,created,release,eol,eol-lts"
#define DISTRO_NAME "Debian"
#define NAME "debian-distro-info"

// C standard libraries
#include <stdlib.h>

#include "distro-info-util.h"

static bool filter_devel(const date_t *date, const distro_t *distro) {
    return created(date, distro) && !released(date, distro) &&
           *distro->version == '\0';
}

static bool filter_oldstable(const date_t *date, const distro_t *distro) {
    return created(date, distro) && released(date, distro);
}

static bool filter_testing(const date_t *date, const distro_t *distro) {
    return created(date, distro) && !released(date, distro);
}

static const distro_t *select_first(const distro_elem_t *distro_list) {
    return distro_list->distro;
}

static const distro_t *select_oldstable(const distro_elem_t *distro_list) {
    const distro_t *newest;
    const distro_t *second = NULL;

    newest = distro_list->distro;
    while(distro_list != NULL) {
        distro_list = distro_list->next;
        if(distro_list) {
            if(date_ge(distro_list->distro->milestones[MILESTONE_RELEASE],
                       newest->milestones[MILESTONE_RELEASE])) {
                second = newest;
                newest = distro_list->distro;
            } else if(second && date_ge(distro_list->distro->milestones[MILESTONE_RELEASE],
                                        second->milestones[MILESTONE_RELEASE])) {
                second = distro_list->distro;
            }
        }
    }
    return second;
}

#include "distro-info-util.c"
